dbms/src/AggregateFunctions/AggregateFunctionAvg.h

```diff
@@ -43,29 +43,31 @@ template <typename T, typename Data>
 class AggregateFunctionAvg final : public IAggregateFunctionDataHelper<Data, AggregateFunctionAvg<T, Data>>
 {
 public:
-    using ResultType = std::conditional_t<IsDecimalNumber<T>, Decimal128, Float64>;
-    using ResultDataType = std::conditional_t<IsDecimalNumber<T>, DataTypeDecimal<Decimal128>, DataTypeNumber<Float64>>;
+    using ResultType = std::conditional_t<IsDecimalNumber<T>, T, Float64>;
+    using ResultDataType = std::conditional_t<IsDecimalNumber<T>, DataTypeDecimal<T>, DataTypeNumber<Float64>>;
     using ColVecType = std::conditional_t<IsDecimalNumber<T>, ColumnDecimal<T>, ColumnVector<T>>;
-    using ColVecResult = std::conditional_t<IsDecimalNumber<T>, ColumnDecimal<Decimal128>, ColumnVector<Float64>>;
+    using ColVecResult = std::conditional_t<IsDecimalNumber<T>, ColumnDecimal<T>, ColumnVector<Float64>>;
 
     /// ctor for native types
     AggregateFunctionAvg(const DataTypes & argument_types_)
         : IAggregateFunctionDataHelper<Data, AggregateFunctionAvg<T, Data>>(argument_types_, {})
         , scale(0)
+        , precision(0)
     {}
 
     /// ctor for Decimals
     AggregateFunctionAvg(const IDataType & data_type, const DataTypes & argument_types_)
         : IAggregateFunctionDataHelper<Data, AggregateFunctionAvg<T, Data>>(argument_types_, {})
         , scale(getDecimalScale(data_type))
+        , precision(getDecimalPrecision(data_type))
```

Comment:
We'd better save old logic here cause it's inconsistent with other places. We have no logic to check real decimal overflows so precision is something misleading. The only one place we check it is String to Decimal conversion.

Reply:
fixed

```
CREATE TABLE test_23
(
    `d1` Decimal(6, 2), 
    `d2` Decimal(11, 4), 
    `d3` Decimal(28, 1)
)
ENGINE = Memory

Ok.

0 rows in set. Elapsed: 0.012 sec. 

akonyaev-laptop :) select toTypeName(avg(d1)), toTypeName(avg(d2)),toTypeName(avg(d3)) from test_23

SELECT 
    toTypeName(avg(d1)), 
    toTypeName(avg(d2)), 
    toTypeName(avg(d3))
FROM test_23

┌─toTypeName(avg(d1))─┬─toTypeName(avg(d2))─┬─toTypeName(avg(d3))─┐
│ Decimal(9, 2)       │ Decimal(18, 4)      │ Decimal(38, 1)      │
└─────────────────────┴─────────────────────┴─────────────────────┘

```

tests fixed